File: programmers/땅따먹기/땅따먹기.py
def get_max_value_except_self_idx(row, idx):
    max_value = 0

    for i in range(len(row)):
        if i == idx:
            continue
        max_value = max(max_value, row[i])

    return max_value


def solution(land):
    n = len(land)
    r = 4
    dp = [[0] * r for _ in range(n)]

    for i in range(r):
        dp[0][i] = land[0][i]

    for i in range(1, n):
        for j in range(r):
            dp[i][j] = get_max_value_except_self_idx(dp[i - 1], j) + land[i][j]

    return max(dp[n - 1])


print(solution([[1, 2, 3, 5], [5, 6, 7, 8], [4, 3, 2, 1]]))

File: programmers/땅따먹기/땅따먹기.java
import java.util.Arrays;

import static java.lang.Math.*;

public class Main {
    public static void main(String[] args){
        Solution solution = new Solution();
        System.out.println(solution.solution(new int[][]{{1, 2, 3, 5}, {5, 6, 7, 8}, {4, 3, 2, 1}}));
    }
}

class Solution {
    int solution(int[][] land) {
        int n = land.length;
        int[][] dp = new int[n][4];
        System.arraycopy(land[0], 0, dp[0], 0, 4);

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < 4; j++) {
                for (int k = 0; k < 4; k++) {
                    if (j == k) {
                        continue;
                    }
                    dp[i][j] = max(dp[i][j], dp[i - 1][k]);
                }
                dp[i][j] += land[i][j];
            }
        }

        return Arrays.stream(dp[n - 1]).max().getAsInt();
    }
}
